docs/01-app/02-guides/self-hosting.mdx

```diff
@@ -267,3 +267,19 @@ if (process.env.NEXT_MANUAL_SIG_HANDLE) {
 ```
 
 </PagesOnly>
+
+## Delay Graceful Shutdowns
+
+Next.js will listen for `SIGINT` and `SIGTERM` and perform a graceful shutdown. If you need to delay handling those signals - e.g. to let your Kubernetes readiness probe report “not ready” before the process exits - you can set `NEXT_DELAY_SIG_HANDLE` to a number of milliseconds.
+
+> **Good to know**: Delaying signal handling is not available in `next dev`.
+
+```json filename="package.json"
+{
+  "scripts": {
+    "dev": "next dev",
+    "build": "next build",
+    "start": "NEXT_DELAY_SIG_HANDLE=10000 next start"
```

Comment:
An arbitrary delay is not ideal and instead we have `NEXT_MANUAL_SIG_HANDLE` which allows registering your own signal handling disabling the default ones.

Reply:
That doesn't help us in the context of Kubernetes though and it won't be arbitrary. It will be based on the configuration of our readiness probes.
Here is a good article describing and visualizing the problem
https://blog.risingstack.com/graceful-shutdown-node-js-kubernetes/

We could use `NEXT_MANUAL_SIG_HANDLE` if there was another way to trigger the graceful shutdown `cleanup` function. Manually calling `process.exit(0)` on `SIGTERM` is no longer a graceful shutdown.

Reply:
Oh, would it be possible to do something like this assuming i have set `NEXT_MANUAL_SIG_HANDLE=true`?

```ts
  process.on('SIGINT', () => {
	// timeout, cleanup...
    delete process.env.NEXT_MANUAL_SIG_HANDLE
    process.emit('SIGINT')
  })
```

Reply:
I guess not, because the event listeners don't exist in that. But if the check was inside the handler it would work, right?

Reply:
@ijjk what if we did this instead to enable manually triggering `cleanup` as described above?

```diff
        const cleanup = () => {
+         // Allow the graceful termination to be manually configurable
+         if (process.env.NEXT_MANUAL_SIG_HANDLE) return
+
          if (cleanupStarted) {
```

```diff
        // Make sure commands gracefully respect termination signals (e.g. from Docker)
-        // Allow the graceful termination to be manually configurable
-        if (!process.env.NEXT_MANUAL_SIG_HANDLE) {
          process.on('SIGINT', cleanup)
          process.on('SIGTERM', cleanup)
-        }
```